### ISSUE TITLE
endless loop in parsing multipart form-data headers

### ISSUE THREAD

**Issue by ben-sol** [yhirose/cpp-httplib#396]
line 2222: if there are remaining characters in the buf_ - in my case there were an additional CR(13 or '\r') character left -, the main loop keeps spinning....
 While it may (and should) be discussed in greater detail, if the content is standard compliant or not, I think, we must avoid the endless loop anyway.
[for a scratch workaround, I'm ignoring _all_ outstanding CR or LFs left, but I'm seriously in doubt about the quality of this solution].
Also please note, that the message was not generated 'by hand', but with _request_, the most popular http communication framework in the javascript world...

**Comment by yhirose** [comment]
@ben-sol, thanks for the feedback. But I think the following code cannot create the endless loop you mentioned:

```cpp
auto pos = buf_.find(crlf_);
while (pos != std::string::npos) {
  ...

  buf_.erase(0, pos + crlf_.size());
  off_ += pos + crlf_.size();
  pos = buf_.find(crlf_);
}
break;
```

Could you give me a small possible example to reproduce your problem? (If you can make a pull request to fix it, that would be the best!)

**Comment by ben-sol** [comment]
As I mentioned: If the _input_ has additional elements in the buffer, you keep spinning in state 2 forever.
 So a non-empty buffer is _after_ this loop (before the break). I changed the code in my fork, but it's not worth pulling it, it must be discussed first.
The best is to try to build a client for you (remember: you can't control the input, right)?

**Comment by ben-sol** [comment]
please put this fragment in an empty directory as 'file.js'
then please install node.js and enter the mentioned directory:
>npm install request
>node file.js <url-to-your-server-here>

`if (process.argv.length < 3 ||
    process.argv[2] == '-h' || process.argv[2] == 'h' ||
    process.argv[2] == '--help' || process.argv[2] == '-help' || process.argv[2] == 'help') {
    console.log('Please enter a valid url like http://<hostname>:<port>/whatever')
    process.exit(1)
}
const url = process.argv[2]

const headers = {
    Connection: 'keep-alive'
}

const formData = {
    name: "name1",
    content: "value1"
}

require('request')
.post({url, headers, formData }, function(err, response, body) {
    if (err) {
        console.log('failed: ', err, 'response:', response, 'body:', body)
    }
    else console.log('success: response:', response, 'body:', body)
})
`

... sorry for the clumsy solution.

**Comment by yhirose** [comment]
@ben-sol, when you work on the client, could you also try to make a unit test with `test_raw_request` method in test/test.cc?

https://github.com/yhirose/cpp-httplib/blob/5b51aa6851d4b03753ec359362b3b6a598b3845e/test/test.cc#L2067-L2068

The purpose of this method is to send **incorrect** HTTP data to a server.
Here is one of examples of how to use it in the unit test suite:

https://github.com/yhirose/cpp-httplib/blob/5b51aa6851d4b03753ec359362b3b6a598b3845e/test/test.cc#L2091-L2099

Or could you at least tell me a raw HTTP data string that causes the infinite loop?
Thanks for your help!

**Comment by ben-sol** [comment]
This was the missing link! I planned to add such a raw test possibility...

**Comment by yhirose** [comment]
@ben-sol, it seems to be busy. Could you give me just the raw HTTP string that causes the problem? I'll take care of it since it's a serious problem. Thanks!

**Comment by yhirose** [comment]
@ben-sol, I'll close it for now, because I can't reproduce it. Please let me know when you are ready to give me the problematic HTTP input. Thanks!

**Comment by yhirose** [comment]
@ben-sol, I merged your comment on #438 here.

> Reasons: state = 2 in the mentioned method has two holes (endless loop):
> 1. small input buffer. If you hit the limit, a partial header remains there.unparsable -> remains in state 2 == endless loop)
> 2. any other characters remaining in the first buffer (CR was in mine) between the fragments. A partial header remains ... (rest as before)
> 
> I git your need for 'examples' but you also may _look_ at the code :)
> Still a promise... again....: example at weekend.

I simply don't have to investigate this problem unless I can easily reproduce it...

> @ben-sol, when you work on the client, could you also try to make a unit test with test_raw_request method in test/test.cc?

> @ben-sol, it seems you are busy. Could you give me just the raw HTTP string that causes the problem? I'll take care of it since it's a serious problem. Thanks!

I'll leave it for a while. Hope you will give me the simple code example to reproduce it.
Thanks for your consideration!

**Comment by ben-sol** [comment]
The first try failed. While I'm able to feed without errors your own multipart-test, the framework cores after modifying the content slightly before getting to my point.
Posting the content is meaningless, Because we have also some issues with the 'Content-Length'.
I have to look at the core first.
<Note> I created a second 'test_raw_request' to execute against your ServerTest instance: so I'm executing exactly the same requests as you do.
I'm perfectly fine with the 'need more information',

**Comment by yhirose** [comment]
@ben-sol, thanks for the update.

**Comment by yhirose** [comment]
I'll remove the label 'bug' unless it can be verified and reproduced in a unit test.

**Comment by yhirose** [comment]
@ben-sol, when you work on verifying the issue, the following pull request from @matvore could be helpful for you.

https://github.com/yhirose/cpp-httplib/pull/457/files

**Comment by yhirose** [comment]
@ben-sol, I found time to look into this issue this evening based on your comments below.

> line 2222: if there are remaining characters in the buf_ - in my case there were an additional CR(13 or '\r') character left -, the main loop keeps spinning....
> While it may (and should) be discussed in greater detail, if the content is standard compliant or not, I think, we must avoid the endless loop anyway.
> [for a scratch workaround, I'm ignoring _all_ outstanding CR or LFs left, but I'm seriously in doubt about the quality of this solution].

> As I mentioned: If the input has additional elements in the buffer, you keep spinning in state 2 forever.

Here is my new test case in test.cc based on your comments by adding extra `\r` codes to each header line in the multipart-data body:

https://github.com/yhirose/cpp-httplib/blob/9562d1c3c1c744e59097bc562c95d1e2eaae3741/test/test.cc#L2389-L2404

But it doesn't cause the infinite loop that you mentioned. Please let me know if you find anything missing to reproduce the issue in my test code.

Personally, I would not like to spend time for this issue anymore unless it is really verified with a unit test. I'll keep it open for another few days, then close it if I still don't get any proof.

Thank you for your understanding.

**Comment by yhirose** [comment]
Close it for now until it can be validated with a unit test case.

**Comment by ben-sol** [comment]
Resolved in the current version (0.7): it returns 'false'.
